src/misc/unicode.cpp

```diff
@@ -0,0 +1,1159 @@
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "dosbox.h"
+
+#include <algorithm>
+#include <fstream>
+#include <map>
+#include <set>
+#include <string>
+#include <vector>
+
+#include "checks.h"
+#include "dos_inc.h"
+#include "string_utils.h"
+
+CHECK_NARROWING();
+
+
+class Grapheme final {
+public:
+    Grapheme() = default;
+    Grapheme(const uint16_t code_point);
+
+    bool IsEmpty() const;
+    bool IsValid() const;
+    bool HasMark() const;
+    uint16_t GetCodePoint() const;
+
+    void Invalidate();
+    void AddMark(const uint16_t code_point);
+    void StripMarks();
+
+    bool operator<(const Grapheme &other) const;
+
+private:
+    // Unicode code point
+    uint16_t code_point = static_cast<uint16_t>(' ');
+    // Combining marks, 0 = none, 1 = U+0300, 2 = U+0301, etc.
+    uint16_t mark_1st = 0;
+    uint16_t mark_2nd = 0;
+
+    bool empty = true;
+    bool valid = true;
+};
+
+using code_page_mapping_t = std::map<Grapheme, uint8_t>; // UTF-8 -> code page
+using code_page_mapping_reverse_t = std::map<uint8_t, Grapheme>;
+
+using config_duplicates_t = std::map<uint16_t, uint16_t>;
+using config_aliases_t    = std::vector<std::pair<uint16_t, uint16_t>>;
+
+typedef struct ConfigMappingEntry {
+    bool valid                          = false;
+    code_page_mapping_reverse_t mapping = {};
+    uint16_t extends_code_page          = 0;
+    std::string extends_dir             = "";
+    std::string extends_file            = "";
+} ConfigMappingEntry;
+
+using config_mappings_t = std::map<uint16_t, ConfigMappingEntry>;
+
+static const std::string file_name_main   = "MAIN.TXT";
+static const std::string file_name_ascii  = "ASCII.TXT";
+static const std::string dir_name_mapping = "mapping";
+
+// Thresholds to tell how the code point is encoded
+constexpr uint8_t threshold_non_ascii = 0x80; // 1000 0000
+constexpr uint8_t threshold_encoded   = 0xc0; // 1100 0000
+constexpr uint8_t threshold_bytes_3   = 0xe0; // 1110 0000
+constexpr uint8_t threshold_bytes_4   = 0xf0; // 1111 0000
+constexpr uint8_t threshold_bytes_5   = 0xf8; // 1111 1000
+constexpr uint8_t threshold_bytes_6   = 0xfc; // 1111 1100
+
+// Use the character below if there is absolutely no sane way to handle UTF-8 glyph
+constexpr uint8_t unknown_character = 0x3f; // '?'
+
+// End of file marking, use in some files from unicode.org
+constexpr uint8_t end_of_file_marking = 0x1a;
+
```

Comment:
Just a suggestion - consider bit literals too, if that's how the documentation describes them.

Reply:
0x1a is a 'magic number'. I have found no documentation regarding this value.
0x3f is simply an ASCII code.